aeppl/loglik.py

```diff
@@ -1,126 +1,173 @@
+    # produced when `Op`s are lifted through them.
+    rv_remapper = PreserveRVMappings(rv_values)
 
-        if rv_value_var is None:
-            rv_value_var = rv_value
+    fgraph.attach_feature(rv_remapper)
 
-    if rv_var is None:
-        if var.owner is not None:
-            return _loglik(
-                var.owner.op,
-                var,
-                rv_values,
-                *var.owner.inputs,
-            )
+    _ = optimize_graph(fgraph, custom_opt=RVSinker())
 
-        return at.zeros_like(var)
+    # This is the updated random-to-value-vars map with the
+    # lifted variables
+    lifted_rv_values = rv_remapper.rv_values
+    replacements = lifted_rv_values.copy()
 
-    rv_node = rv_var.owner
+    # Walk the graph from its inputs to its outputs and construct the
+    # log-likelihood
+    q = deque(fgraph.toposort())
 
-    rng, size, dtype, *dist_params = rv_node.inputs
+    logpdf_var = None
 
-    # Here, we plug the actual random variable into the log-likelihood graph,
-    # because we want a log-likelihood graph that only contains
-    # random variables.  This is important, because a random variable's
-    # parameters can contain random variables themselves.
-    # Ultimately, with a graph containing only random variables and
-    # "deterministics", we can simply replace all the random variables with
-    # their value variables and be done.
+    while q:
+        node = q.popleft()
 
-    # tmp_rv_values = rv_values.copy()
-    # tmp_rv_values[rv_var] = rv_var
+        if not any(o in lifted_rv_values for o in node.outputs):
+            if isinstance(node.op, RandomVariable) and warn_missing_rvs:
+                warnings.warn(
+                    "Found a random variable that was neither among the observations "
+                    f"nor the conditioned variables: {node}"
+                )
+            continue
 
-    logpdf_var = _logpdf(rv_node.op, rv_value_var, *dist_params, **kwargs)
+        if isinstance(node.op, RandomVariable):
+            q_rv_var = node.outputs[1]
+            q_rv_value_var = replacements[q_rv_var]
 
-    # Replace random variables with their value variables
-    replacements = rv_values.copy()
-    replacements.update({rv_var: rv_value, rv_value_var: rv_value})
+            # Replace `RandomVariable`s in the inputs with value variables.
+            # Also, store the results in the `replacements` map so that we
+            # don't need to redo these replacements.
+            value_var_inputs, _ = rvs_to_value_vars(
+                node.inputs,
+                initial_replacements=replacements,
+            )
 
-    (logpdf_var,), _ = rvs_to_value_vars(
-        (logpdf_var,),
-        initial_replacements=replacements,
-    )
+            q_logpdf_var = _logpdf(node.op, q_rv_value_var, *value_var_inputs, **kwargs)
 
-    if sum:
-        logpdf_var = at.sum(logpdf_var)
+        else:
+            raise NotImplementedError(
```

Comment:
Is the plan to dispatch to other Ops in here (such as for the affine transform)?

Reply:
No, that's all handled via rewrites in `RVSinker`.

Reply:
Okay I think I see what you have in mind. Is the goal to convert all graphs into basic random-variable nodes first? Sounds a bit more limited than what I had in mind, but perhaps it's the best first approach.

Reply:
> Sounds a bit more limited than what I had in mind, but perhaps it's the best first approach.

It's not limited in any way&mdash;at least not relative to the dispatch approach.  It is significantly simpler, though.

For example, I just finished implementing everything, and, now, `joint_logprob` is very direct and shouldn't need to be changed in order to provide any of the features we've considered (e.g. support for convolutions, affine transforms, `Scan`/`for`-loops, etc.)

In the present form, `joint_logprob` takes a graph, rewrites it, and forms the sum of each `RandomVariable`'s log-probability.  That's it.

This neatly compartmentalizes the rewriting work by placing it squarely within the optimization/rewriting framework where it belongs.